apps/sms/views/shared/js/navigation.js

```diff
@@ -1,361 +1,726 @@
    * - nextPanel.afterEnter
+   * - previousPanel.afterLeave
    *
-   * The panels are slid between beforeEnter and afterLeave.
+   * If we stay in the same document and enter a new panel, the panels are slid
+   * between beforeEnter and afterLeave.
    *
-   * Therefore it is expected that:
-   * - beforeLeave does not change any state but merely checks whether the
-   *   transition can take place
-   * - beforeEnter checks whether the transition can take place and make any
-   *   visual change that should be made before the transition
-   * - afterLeave cleans up the panel, reclaim memory
+   * With the new architecture, the navigation can happen in one of 2 cases:
+   * 1. either we stay in the same document,
+   * 2. or we move to another document.
+   * 
+   * In all cases:
+   * - beforeLeave can prevent changing the location and thus the
+   *   navigation. It is expected that beforeLeave does not change any state but
+   *   merely checks whether the transition can take place.
+   * - beforeEnter will be called before the view is displayed. Therefore it can
+   *   make visual changes but it can't prevent a view from being entered. All
+   *   rejections and exceptions will be swallowed and output to the console.
+   * - afterLeave needs to clean up the panel, reclaim memory
    * - afterEnter does left-over actions
    *
-   * beforeLeave and beforeEnter can :
-   * - throw or return a rejected Promise to stop the transition.
-   * - finish normally or return a resolved Promise to continue the transition
-   *
-   * However, afterLeave and afterEnter should only finish normally or return a
-   * resolved Promise. Otherwise, the errors will be logged but the transition
-   * will finish.
-   *
    * toPanel itself returns a Promise that will be resolved if the transition
    * happened and rejected otherwise.
+   *
+   * @param {String} viewName The view to load.
+   * @param {Object} [args] The arguments to pass to the view.
+   * @returns {Promise} Resolves when the transition is over.
    */
-  toPanel: function n_toPanel(panel, args) {
-    if (!(panel in this.panels)) {
-      return Promise.reject(new Error('Panel ' + panel + ' is unknown.'));
-    }
+  toPanel(viewName, args) {
+    debug('toPanel(', viewName, ',', args, ')');
+    var view = findViewFromName(viewName);
 
-    // put the request to a queue when:
-    //   - Panel is still transitioning
-    //   - trying to navigate when the app is not loaded completely
-    var notReadyNavigate = (panel !== this.defaultPanel && !this.isReady);
-    if (this.transitionPromise || notReadyNavigate) {
-      queuedPanel = {
-        panel: panel,
-        args: args
-      };
-
-      return new Promise(function(resolve, reject) {
-        queuedPanel.defer = {
-          resolve: resolve,
-          reject: reject
-        };
-      });
-    }
-
-    // this checks both panel and args
-    if (this.isCurrentPanel(panel, args)) {
+    if (view === currentState.panel) {
```

Comment:
I'd say yes (because init is supposed to set it) but maybe we can call toPanel before init if things happen in a specific order at startup; I can make this stricter.